src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/AutoInsert/AutoClosingTagOnAutoInsertProvider.cs

```diff
@@ -39,55 +37,40 @@ internal sealed class AutoClosingTagOnAutoInsertProvider(RazorLSPOptionsMonitor
 
     private static readonly ImmutableHashSet<string> s_voidElementsCaseSensitive = s_voidElements.WithComparer(StringComparer.Ordinal);
 
-    private readonly RazorLSPOptionsMonitor _optionsMonitor = optionsMonitor;
-
     public string TriggerCharacter => ">";
 
-    public bool TryResolveInsertion(Position position, FormattingContext context, [NotNullWhen(true)] out TextEdit? edit, out InsertTextFormat format)
+    public InsertTextEdit? TryResolveInsertion(Position position, RazorCodeDocument codeDocument, bool enableAutoClosingTags)
```

Comment:
I did have this service converted to async temporarily, which is why I switched parameters initially. I didn't see a need to convert it back to bool and out when I removed the async per PR suggestion. I don't mind changing it back to bool, but honest question - why? We had a discussion on the former WebTools before about whether TryXXX methods should return bool and "out" now that there are nullable types (both reference and value) and came to conclusion that returning a nullable type instead and checking for null made more sense.

If you have strong preference for bool and out I will change it, though I honestly see no reason to (I could be missing something of course).

Reply:
So interestingly my answer for "why" is the same as WebTools, I just think "now that there are nullable types" is the reason _to_ prefer the other form. I find it much nicer to write things like:

```
if (!thing.TryResolveInsertion(0, out var edit))
{
	return;
}
```

Now the compiler knows for sure that edit will never be null, and we don't have to think about the fact that there was ever a nullable type in play.

Reply:
Especially since async method can't have out parameters... Seems it would be more consistent to prefer *not* to use out parameters, no?

Reply:
Since Dustin is also in favor of changing it back to bool, I'm overruled and I'll just change it back to being bool.

Reply:
> That requires you to annotate the out parameter with [NotNullWhenTrue] though, right? Wouldn't it be clearer (and simpler) to just do
> 
> ```
> if (thing.TryResolveInsertion(0) is not { } edit))
> {
> 	return;
> }
> ```
> 
> when returning a nullable type instead of bool? If returning null via `out` and `false` via return is equivalent, why do it both ways - returning the same result twice in a single method call?
> 
> Not trying to be argumentative, and honestly don't have that much of a preference one way or the other, just trying to understand the reasoning :)

I don't mind that pattern at all! Although, there are plenty of cases in the PR you weren't using that pattern. 😛 

I disagree that pattern is cleaner and simpler though. When reading your example, I don't get a `!` up front, so I have to read further to see that it's actually a negative test when I get to the `not`. Than, I have to remind myself that `{ }` is really matching an empty property pattern, which is a strange construction in itself that means that just it matches an object instance of an inferred type. Finally, the construction `not { } edit` gives the thing that it "not { }" a name! If one thinks about this pattern too much, it can become kind of twisted!

To me, boolean logic is more readable if I don't have eye-parse and convert extra null/not-null to true/false.  Rewriting your example code, I find this much more readable:

```C#
if (!thing.TryResolveInsertion(0, out var edit))
{
    return;
}
```

So, I'm personally not opposed to the pattern you wrote, but I use it as a last resort and will aim to make `Try` methods return a `bool` because that's a better signal for false/true than null/not-null.

Just my two cents.

Reply:
That's fair, I still find the "is { }" and such hard to read and have to parse it. I don't mind going with whatever the majority finds more readable.